src/actors/workers/blocks.rs

```diff
@@ -85,27 +91,57 @@ where
         Ok(blocks)
     }
 
+    /// Collect blocks according to the predicate `fun` and send those blocks to
+    ///  the metadata actor.
+    async fn collect_and_send(&self, fun: impl Fn(u32) -> bool + Send + 'static) -> Result<()> {
+        self.meta
+            .send(BatchBlock::new(self.collect_blocks(fun).await?))
+            .await?;
+        Ok(())
+    }
+
     /// First run of indexing
     /// gets any blocks that are missing from database and indexes those.
     /// sets the `last_max` value.
-    async fn re_index(&mut self) -> Result<Option<Vec<Block<B>>>> {
+    async fn re_index(&mut self) -> Result<()> {
         let mut conn = self.db.send(GetState::Conn.into()).await?.await?.conn();
-        let numbers = queries::missing_blocks_min_max(&mut conn, self.last_max).await?;
+        let mut numbers = queries::missing_blocks_min_max(&mut conn, self.last_max).await?;
         let len = numbers.len();
         log::info!("{} missing blocks", len);
         self.last_max = if let Some(m) = queries::max_block(&mut conn).await? {
             m
         } else {
             // a `None` means that the blocks table is not populated yet
-            return Ok(None);
+            return Ok(());
         };
-        let blocks = self.collect_blocks(move |n| numbers.contains(&n)).await?;
-        Ok(Some(blocks))
+
+        if numbers.len() as u32 > self.max_block_load {
+            let mut tmp_set = HashSet::<u32>::with_capacity(self.max_block_load as usize);
+            for height in numbers.drain() {
+                tmp_set.insert(height);
+                tmp_set = if tmp_set.len() as u32 > self.max_block_load  {
```

Comment:
What's the rationale behind using HashSet here? It would filter out duplicates however we shouldn't get duplicates unless something is wrong with our SQL query

Rather than use drain/HashSet here, we could iterate using the `chunks` method defined on slices and check if the number is contained within that chunk since we don't need to worry about passing around an owned `height`. This should remove the need to instantiate a new HashSet on every iteration

Reply:
About to try this out, but I would think `chunks().into_iter()` would return owned chunks, and from my understanding this consumes `chunks()` and just change the way its referenced but not touch the heap. 

Also I was trying to figure out yesterday, does `to_vec()` just changed what is stored on the stack or does it also affect the heap?

Reply:
I ended up trying out a solution that modifies the SQL query. From what I read, using LIMIT is said to reduce memory usage of the query and SQL can do some optimizations under the hood. As I am typing right now realizing though that I haven't checked if awaiting the metadata actor is enough to make sure the current batch is inserted - if not this code is not correct.

I tried out solutions with BTreeSet and Vec as well, but this SQL query might be more space and time efficient than the Vec solution by some constants. (I couldn't figure out a a BtreeSet solution with better than nlogn time)

My main issue right now is I can't figure out how to test the modified query because I can't get the database to have missing entries. I kept on trying to just kill the processes fro archive while syncing, but on restart the missing blocks query always comes back empty. Is there an easy way I could seed a DB with just one block at high height?

Reply:
> I was trying to figure out yesterday, does `to_vec()` just change what is stored on the stack or does it also affect the heap?

I am pretty sure `to_vec()` reallocates and clone the data in this case.

Reply:
> I ended up trying out a solution that modifies the SQL query.

Sounds like the most straight forward approach.

> Is there an easy way I could seed a DB with just one block at high height?

Why not just insert a new record? Clone a row of data and manually edit the height column?

Reply:
@emostov just a sidenote, I've sometimes used [`alloc_counter`](https://lib.rs/crates/alloc_counter) when I had doubt about how much a piece of code allocates. It's far from perfect as it can't see any allocations that happens outside the Rust code (e.g. Postgres or Rocksdb) but still handy at times, e.g. in benchmarks or while debugging something.

Reply:
> Is there an easy way I could seed a DB with just one block at high height?

If you install the SQLx CLI: https://github.com/launchbadge/sqlx/tree/master/sqlx-cli you should be able to drop the current database  and create a new one, then run the migrations for substrate-archive with `cargo sqlx migrate run` after setting the DATABASE_URL environment variable.

Then, this query should insert block 101000 (should work through psql):
```sql
INSERT INTO blocks (parent_hash, hash, block_num, state_root, extrinsics_root, digest, ext, spec) VALUES (E'\\x8da3783495634a1645cddb0d5b72743170971a97b68391ce2eb4cbe6a1abebcc', E'\\xa8e323dce4492818f6ae828ea0e483c572536ec8b56a742c351b25fccc0c428f', 101000, E'\\x407c1a180a91c3fed506ac7a1695b31bc830e46c98a8d8397430f2f0bc8f79b0', E'\\xfebdb8c5a254c0e58831b1d01d398abd74979a5ec468797ef9a3b80e9f5d4fb0', E'\\x08064241424534023b0000001eeea60f0000000005424142450101b406381a5d732324723a54ee0951eecf43b621af610ce844fc297445e270de416311da7ef66300f8d637fc9ea907e19478ebcc9307589a376ba05215ca582d82', E'\\x0c280402000b20df6cd86e011c0409001a2a06001004140000', 1028);
```

> I haven't checked if awaiting the metadata actor is enough to make sure the current batch is inserted - if not this code is not correct.

I just checked up on this and you're right, it won't, but it should be a simple fix. Adding another `.await` here:
https://github.com/paritytech/substrate-archive/blob/70c6d9d3bad8cbfb5ac7ed407eb8acc04fbf26d3/src/actors/workers/metadata.rs#L88

Is all it should take to fix, the first future returns a future that can be used to wait for the result of the first future (not really intuitive, but that api is out-of-scope for this PR I think)